### PR TITLE
[DOC] Codebase walkthrough with Vector add example

### PR DESCRIPTION
**Pull request by masahi**
The first stab at #2160.

@tqchen @yzhliu @merrymercy @Ravenwater please review.

### CODE REVIEW

File: docs/dev/codebase_walkthrough.rst [hunk 1]
```diff
@@ -0,0 +1,227 @@
...
+::
+
+   target = "cuda"
+   fadd = tvm.build(s, [A, B, C], target)
+
+``tvm.build()``, defined in ``python/tvm/build_module.py``, takes a schedule, input and output ``Tensor``, and a target, and returns a ``tvm.Module`` object, defined in ``python/tvm/module.py``. A ``Module`` object contains a compiled function which can be invoked with function call syntax.
+
+The process of ``tvm.build()`` can be divided into two steps:
+
+- Lowering, where an high level, initial loop nest structures are transformed into a final, low level IR
```
Comment (vinx13) [inline docs/dev/codebase_walkthrough.rst]: a high level

File: docs/dev/codebase_walkthrough.rst [hunk 2]
```diff
@@ -0,0 +1,227 @@
...
+
+When an user invokes graph compilation by ``nnvm.compiler.build(...)``, the following sequence of actions happens for each node in the graph:
+
+- Look up an operator implementation by querying the operator registry
+- Generate a compute expression and a schdule for the operator
+- Compile the operator into object code
+
+One of the interesting aspects of TVM codebase is that interop between C++ and Python is not unidirectional. Typically, all code that do heavy liftings are implemented in C++, and Python bindings are provided for user interface. This is also true in TVM, but in TVM codebase, C++ code also call into functions defined in a Python module. For example, the convolution operator is implemented in Python, and its implementation is invoked from C++ code in nnvm.
+
+At the time of writing (Nov. 30, 2018), there is an going effort to reimplement functionality offered by ``nnvm`` in a new intermidiate representation called Relay. New Relay code resides in ``src/relay`` and ``python/tvm/relay``.
```
Comment (vinx13) [inline docs/dev/codebase_walkthrough.rst]: an ongoing effort

File: docs/dev/codebase_walkthrough.rst [hunk 3]
```diff
@@ -0,0 +1,227 @@
...
+We use a simple example that does not use topi or nnvm. The example is vector addition, which is covered in detail in `this tutorial <https://docs.tvm.ai/tutorials/get_started.html#sphx-glr-tutorials-get-started-py>`_.
+
+::
+
+   n = 1024
+   A = tvm.placeholder((n,), name='A')
+   B = tvm.placeholder((n,), name='B')
+   C = tvm.compute(A.shape, lambda i: A[i] + B[i], name="C")
+
+Here, types of ``A``, ``B``, ``C`` are ``tvm.tensor.Tensor``, defined in ``python/tvm/tensor.py``. The Python ``Tensor`` is backed by C++ ``Tensor``, implemented in ``include/tvm/tensor.h`` and ``src/lang/tensor.cc``. All Python types in TVM can be thought of as a handle to the underlining C++ type with the same name. If you look at the definition of Python ``Tensor`` type below, you can see it is an subclass of ``NodeBase``.
```
Comment (vinx13) [inline docs/dev/codebase_walkthrough.rst]: an subclass -> a subclass

File: docs/dev/codebase_walkthrough.rst (hunk 2 above)
Comment (vinx13) [reply docs/dev/codebase_walkthrough.rst]: intermidiate -> intermediate

**Comment by masahi** [comment]
@vinx13 Thanks. Fixed.

**Comment by jroesch** [comment]
Would probably be good to port this to Relay soonish too. One of the Relay developers can probably manage that though, I will add a note to the tracking issue.

File: docs/dev/codebase_walkthrough.rst [hunk 4]
```diff
@@ -0,0 +1,227 @@
...
+We use a simple example that does not use topi or nnvm. The example is vector addition, which is covered in detail in `this tutorial <https://docs.tvm.ai/tutorials/get_started.html#sphx-glr-tutorials-get-started-py>`_.
+
+::
+
+   n = 1024
+   A = tvm.placeholder((n,), name='A')
+   B = tvm.placeholder((n,), name='B')
+   C = tvm.compute(A.shape, lambda i: A[i] + B[i], name="C")
+
+Here, types of ``A``, ``B``, ``C`` are ``tvm.tensor.Tensor``, defined in ``python/tvm/tensor.py``. The Python ``Tensor`` is backed by C++ ``Tensor``, implemented in ``include/tvm/tensor.h`` and ``src/lang/tensor.cc``. All Python types in TVM can be thought of as a handle to the underlining C++ type with the same name. If you look at the definition of Python ``Tensor`` type below, you can see it is a subclass of ``NodeBase``.
```
Comment (merrymercy) [inline docs/dev/codebase_walkthrough.rst]: `underlining` -> `underlying`

**Pull request closed without merging**